### PR TITLE
fix leak in hare outputs collection

### LINKED ISSUE #1597: hare doesn't empty hare result buffer
# Overview / Motivation

in `hare.go` when running `collectOutput` it tries to checks whether the `outputs` map size is equal to `bufferSize` if it does then it iterates over the outputs and delete any of them who returns true for `IsTooLate`, which checks whether `inputlayerid < lastLayer - oldestInBuffer`
so let's say the buffer size is `20`,  we got output for layer `21`, we iterate over the outputs map and check whether layer 4 (for example) is too late. `4 < 21 - 1 == false` this will be false, layer 21 will be added to the buffer without deleteing any item and next iteration the condition to check this (if the outputs size is equal to `bufferSize`) will be false forever and the buffer keeps growing for ever.

# The Task

Change the code so the buffer get's emptied when new outputs are arriving.

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who haven't introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### PR DESCRIPTION
**Pull request by y0sher**
fixes #1597

### CODE REVIEW

File: hare/hare.go [hunk 1]
```diff
@@ -119,23 +119,31 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
...
+	if id < instanceId(lyr-types.LayerID(h.bufferSize)) { // bufferSize>=0
 		return true
 	}
 	return false
 }
 
 func (h *Hare) oldestResultInBuffer() types.LayerID {
-
+	// buffer is usually quite small so its cheap to iterate.
+	// TODO: if it gets bigger change `outputs` to array.
```
Comment (barakshani) [inline hare/hare.go:143]: Please add a comment that the buffer size should be at least hdist

File: hare/hare.go [hunk 2]
```diff
@@ -119,23 +119,31 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
 
 // checks if the provided id is too late/old to be requested.
 func (h *Hare) isTooLate(id instanceId) bool {
```
Comment (barakshani) [inline hare/hare.go]: we can name this better... e.g. includeInOutputBuffer

File: hare/hare.go [hunk 3]
```diff
@@ -164,12 +172,8 @@ func (h *Hare) collectOutput(output TerminationOutput) error {
...
-	if len(h.outputs) == h.bufferSize {
-		for k := range h.outputs {
-			if h.isTooLate(instanceId(k)) {
-				delete(h.outputs, k)
-			}
-		}
+	if len(h.outputs) >= h.bufferSize {
+		delete(h.outputs, h.oldestResultInBuffer())
 	}
 	h.outputs[types.LayerID(id)] = blocks
```
Comment (barakshani) [inline hare/hare.go:181]: shouldn't we ask if it's too late before we add to buffer?
(I'm not sure that we should)

**Review by barakshani — APPROVED** [review]

File: hare/hare.go (hunk 2 above)
Comment (y0sher) [reply hare/hare.go]: currently, the new impl doesn't check if it is actually inside the buffer, but checks if the given id is more than `bufferSize` in the past.

File: hare/hare.go (hunk 3 above)
Comment (y0sher) [reply hare/hare.go:181]: if you look a few lines above you'll see that we're checking it beforehand.

File: hare/hare.go (hunk 2 above)
Comment (barakshani) [reply hare/hare.go]: I meant _(should we) include in hare output buffer_ so the returned answer should be true/false
But other names are welcome. `IsTooLate` has no context.

File: hare/hare.go [hunk 4]
```diff
@@ -119,23 +119,31 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
...
+	// TODO: if it gets bigger change `outputs` to array.
 	h.layerLock.RLock()
-
-	if h.lastLayer <= types.LayerID(h.bufferSize) {
-		h.layerLock.RUnlock()
-		return 0
-	}
-
-	lyr := h.lastLayer - types.LayerID(h.bufferSize)
+	lyr := h.lastLayer
```
Comment (antonlerner) [inline hare/hare.go]: I'd refactor this to a getter method

File: hare/hare.go (hunk 2 above)
Comment (y0sher) [reply hare/hare.go]: how about, `outOfHdist` ?

Comment (barakshani) [reply hare/hare.go]: it is not really correlated to hdist at the moment.

**Pull request merged**